addons/mail/models/mail_message.py

```diff
@@ -825,56 +808,78 @@ def unstar_all(self):
-            partner.starred_message_ids -= self
-        self.env.user._bus_send(
-            "mail.message/toggle_star", {"message_ids": [self.id], "starred": starred}
+        if starred := self.filtered("starred"):
+            self.env.user.partner_id.starred_message_ids -= starred
+            self.env.user._bus_send(
+                "mail.message/toggle_star", {"message_ids": starred.ids, "starred": False}
+            )
+        if unstarred := self - starred:
+            self.env.user.partner_id.starred_message_ids |= unstarred
+            self.env.user._bus_send(
+                "mail.message/toggle_star", {"message_ids": unstarred.ids, "starred": True}
         )
 
+    @api.model
+    def _message_fetch(self, domain, search_term=None, before=None, after=None, around=None, limit=30):
+        res = {}
+        if search_term:
+            # we replace every space by a % to avoid hard spacing matching
+            search_term = search_term.replace(" ", "%")
+            domain = expression.AND([domain, expression.OR([
+                # sudo: access to attachment is allowed if you have access to the parent model
+                [("attachment_ids", "in", self.env["ir.attachment"].sudo()._search([("name", "ilike", search_term)]))],
+                [("body", "ilike", search_term)],
+                [("subject", "ilike", search_term)],
+                [("subtype_id.description", "ilike", search_term)],
+            ])])
+            domain = expression.AND([domain, [("message_type", "not in", ["user_notification", "notification"])]])
+            res["count"] = self.search_count(domain)
+        if around is not None:
+            messages_before = self.search(domain=[*domain, ('id', '<=', around)], limit=limit // 2, order="id DESC")
+            messages_after = self.search(domain=[*domain, ('id', '>', around)], limit=limit // 2, order='id ASC')
+            return {**res, "messages": (messages_after + messages_before).sorted('id', reverse=True)}
+        if before:
+            domain = expression.AND([domain, [('id', '<', before)]])
+        if after:
+            domain = expression.AND([domain, [('id', '>', after)]])
+        res["messages"] = self.search(domain, limit=limit, order='id ASC' if after else 'id DESC')
+        if after:
+            res["messages"] = res["messages"].sorted('id', reverse=True)
+        return res
+
     def _message_reaction(self, content, action):
         self.ensure_one()
         partner, guest = self.env["res.partner"]._get_current_persona()
         # search for existing reaction
-        domain = [
-            ("message_id", "=", self.id),
-            ("partner_id", "=", partner.id),
-            ("guest_id", "=", guest.id),
-            ("content", "=", content),
-        ]
-        reaction = self.env["mail.message.reaction"].search(domain)
+        reaction = self.reaction_ids.filtered(lambda r:
+            r.partner_id == partner.id and r.guest_id == guest.id and r.content == content
```

Comment:
```suggestion
            r.partner_id == partner and r.guest_id == guest and r.content == content
```
partner_id is a model instance, compare it with another instance instead of int